### ISSUE TITLE
Gradle scan link is gone

### ISSUE THREAD

**Issue by alaershov** [cdsap/Talaiot#276]
Hello! 

Noticed that you've dropped the Gradle Build Scan link recently, can we please have it back? It was nice to have this link in our Graphana dashboard next to each build.

Or do you suggest we should implement it ourselves and send it as a custom metric?

**Comment by cdsap** [comment]
Hi @alaershov, the feature was removed here https://github.com/cdsap/Talaiot/commit/3569daea16680a56d1ceccd90d39bb6c1dbe7fea because the original implementation `org.gradle.internal.scan.eob.DefaultBuildScanEndOfBuildNotifier` was moved(or removed).
Didn't check if we can have It back, but definitely I leave  this ticket open to research more

**Comment by cdsap** [comment]
btw, @Malinskiy do you know if Gradle is still supporting the DefaultBuildScanEndOfBuildNotifier in Gradle 6.x versions?

**Comment by Malinskiy** [comment]
@cdsap 

There is a new way to get the publisher link via a proper callback. I've implemented a skeleton in https://github.com/cdsap/Talaiot/blob/feature/gradle-scan-link/library/core/talaiot/src/main/kotlin/com/cdsap/talaiot/metrics/GradleScanLinkMetric.kt#L15

This prints the metric to stdout after you accept the license agreement. Unfortunately, the lifecycle problem is coming back again: gradle scan lifecycle is outside of `org.gradle.BuildListener` and we have to send events later when the scan is finished. In the previous implementation I've called the producer part of the link manually and hence the problem was masked.

Here is the stack when the callback gets executed:
```
java.lang.Throwable
        at java.base/java.lang.Thread.dumpStack(Thread.java:511)
        at com.cdsap.talaiot.metrics.GradleScanLinkListener.execute(GradleScanLinkMetric.kt:30)
        at com.cdsap.talaiot.metrics.GradleScanLinkListener.execute(GradleScanLinkMetric.kt:27)
        at com.gradle.enterprise.gradleplugin.internal.extension.b$5.a(SourceFile:160)
        at com.gradle.enterprise.gradleplugin.internal.extension.b$5.accept(SourceFile:157)
        at com.gradle.scan.plugin.internal.o.b.b.a(SourceFile:45)
        at com.gradle.scan.plugin.internal.o.b.b$$Lambda$1149/0x0000000000000000.accept(Unknown Source)
        at java.base/java.lang.Iterable.forEach(Iterable.java:75)
        at com.gradle.scan.plugin.internal.o.b.b.a(SourceFile:43)
        at com.gradle.scan.plugin.internal.n.k.a(SourceFile:318)
        at com.gradle.scan.plugin.internal.n.k$$Lambda$982/0x0000000000000000.accept(Unknown Source)
        at com.gradle.scan.plugin.internal.k.b.a(SourceFile:277)
        at com.gradle.scan.plugin.internal.k.b.a(SourceFile:180)
        at com.gradle.scan.plugin.internal.k.b.a(SourceFile:128)
        at com.gradle.scan.plugin.internal.f.c.d.a(SourceFile:76)
        at com.gradle.scan.plugin.internal.n.b.a(SourceFile:73)
        at com.gradle.scan.plugin.internal.c.d$c.a(SourceFile:60)
        at com.gradle.scan.plugin.internal.c.d$c$$Lambda$1143/0x0000000000000000.buildFinished(Unknown Source)
        at org.gradle.internal.enterprise.impl.DefaultGradleEnterprisePluginAdapter.buildFinished(DefaultGradleEnterprisePluginAdapter.java:77)
        at org.gradle.internal.enterprise.core.GradleEnterprisePluginManager.buildFinished(GradleEnterprisePluginManager.java:73)
        at org.gradle.launcher.exec.BuildCompletionNotifyingBuildActionRunner.run(BuildCompletionNotifyingBuildActionRunner.java:46)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner$3.call(RunAsBuildOperationBuildActionRunner.java:49)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner$3.call(RunAsBuildOperationBuildActionRunner.java:44)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$$Lambda$135/0x0000000000000000.apply(Unknown Source)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner.run(RunAsBuildOperationBuildActionRunner.java:44)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.lambda$execute$0(InProcessBuildActionExecuter.java:54)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter$$Lambda$134/0x0000000000000000.transform(Unknown Source)
        at org.gradle.composite.internal.DefaultRootBuildState.run(DefaultRootBuildState.java:87)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:53)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:29)
        at org.gradle.launcher.exec.BuildTreeScopeLifecycleBuildActionExecuter.lambda$execute$0(BuildTreeScopeLifecycleBuildActionExecuter.java:33)
        at org.gradle.launcher.exec.BuildTreeScopeLifecycleBuildActionExecuter$$Lambda$81/0x0000000000000000.apply(Unknown Source)
        at org.gradle.internal.buildtree.BuildTreeState.run(BuildTreeState.java:49)
        at org.gradle.launcher.exec.BuildTreeScopeLifecycleBuildActionExecuter.execute(BuildTreeScopeLifecycleBuildActionExecuter.java:32)
        at org.gradle.launcher.exec.BuildTreeScopeLifecycleBuildActionExecuter.execute(BuildTreeScopeLifecycleBuildActionExecuter.java:27)
        at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:104)
        at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:55)
        at org.gradle.tooling.internal.provider.SubscribableBuildActionExecuter.execute(SubscribableBuildActionExecuter.java:64)
        at org.gradle.tooling.internal.provider.SubscribableBuildActionExecuter.execute(SubscribableBuildActionExecuter.java:37)
        at org.gradle.tooling.internal.provider.SessionScopeLifecycleBuildActionExecuter.lambda$execute$0(SessionScopeLifecycleBuildActionExecuter.java:54)
        at org.gradle.tooling.internal.provider.SessionScopeLifecycleBuildActionExecuter$$Lambda$74/0x0000000000000000.apply(Unknown Source)
        at org.gradle.internal.session.BuildSessionState.run(BuildSessionState.java:67)
        at org.gradle.tooling.internal.provider.SessionScopeLifecycleBuildActionExecuter.execute(SessionScopeLifecycleBuildActionExecuter.java:50)
        at org.gradle.tooling.internal.provider.SessionScopeLifecycleBuildActionExecuter.execute(SessionScopeLifecycleBuildActionExecuter.java:36)
        at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:36)
        at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:25)
        at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:59)
        at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:31)
        at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:55)
        at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:41)
        at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:47)
        at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:31)
        at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:65)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:39)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:29)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:35)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.create(ForwardClientInput.java:78)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.create(ForwardClientInput.java:75)
        at org.gradle.util.Swapper.swap(Swapper.java:38)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:75)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.LogAndCheckHealth.execute(LogAndCheckHealth.java:55)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:63)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:84)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:52)
        at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:297)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
        at java.base/java.lang.Thread.run(Thread.java:836)
```

The only place I could find to do the same trick is `GradleEnterprisePluginManager`force call the enterprise listener, then reset the listener to null.
From the source code of gradle `DefaultBuildScanEndOfBuildNotifier` has been removed in 6.6.0, which makes our old approach work up to 6.5.1.

I'm not sure if we want to support old implementation or just ask people to use old plugin for older versions of gradle.

I didn't write any tests for the code due to limited time and I assume someone will pick it up.

**Comment by cdsap** [comment]
Thank you very much for the complete description of the problem. 
It is indeed more elegant the solution with  `GradleEnterprisePluginManager`.
I'm going to review the skeleton and try to understand better the problems with the callbacks.

In terms of support, Gradle < 6 was dropped since 1.4.1. The motivation behind `InternalBuildListener` is to support the configuration-cache feature. `BuildListener` was breaking the compatible configuration cache methods.

Thanks again. I release 1.4.2 this week and later will continue with this issue.

**Comment by cdsap** [comment]
Fixed by #278 
thanks @Malinskiy
